data-collection/src/main/java/refactoringml/db/Database.java

```diff
@@ -74,8 +78,29 @@ public long findAllStableCommits(long projectId, int level) {
 		return Long.parseLong(result.toString());
 	}
 
-	public void rollback() {
-		session.getTransaction().rollback();
+	//safely rollback a transaction with the db
+	public void rollback() throws SQLException {
+		//this session object itself should never be null, thus we don't check for it
+		//nothing to do in this case, recovering the session or transaction is to much effort and the db takes care of a failed transaction
+		if(!session.isOpen()) {
+			log.error("Encountered is closed session during rollback. Doing nothing.");
+			return;
+		}
+
+		if(!session.isConnected()){
+			Connection connection =	sf.getSessionFactoryOptions().getServiceRegistry().
```

Comment:
that's magic! I didn't know that before, but I suppose you tried and it worked?

Reply:
It works under some conditions